### ISSUE TITLE
Ability to add images to diaspora* post

### ISSUE THREAD

**Issue by noplanman** [DiasPHPora/wp-to-diaspora#124]
Have an extra field to choose if the featured image (or a custom list of images) should be added to the post as uploaded images (not just inline images).
This also needs to be added to the API.

**Comment by manu-p** [comment]
Hi, yes that would be great! At least the featured image for a beginning, just as for the mailpoet newsletter.

Anyway, thanks for the extension!

**Comment by noplanman** [comment]
You can do that with the [title filter](https://github.com/gutobenn/wp-to-diaspora/wiki/Filters#title-filter) :+1:

You can make use of [`get_the_post_thumbnail`](https://developer.wordpress.org/reference/functions/get_the_post_thumbnail/).

Try something like this to start with:
```php
add_filter( 'wp2d_title_filter',
	function( $default, $title, $permalink ) {
		return $default . '<br />' . get_the_post_thumbnail();
	}, 10, 3
);
```

If you need any further help, just ask.

**Comment by manu-p** [comment]
Hi noplanman, I need more explanation, : where am I supposed to paste these code lines ?

Thanks again.

**Comment by noplanman** [comment]
Sure thing, I'll explain a bit further.

Regarding general information about action hooks, read [here](https://codex.wordpress.org/Plugin_API#Actions).

You can drop the code I pasted above into your theme's [`functions.php`](https://codex.wordpress.org/Functions_File_Explained) file. Anything in that file gets executed when the theme loads.

I suggest you add the code and test posting something to diaspora*. Then, you can modify some things and create a new post to see it in action.

Are you familiar with PHP and HTML code?
Basically, in the function from the code I posted above, you can simply return any HTML content and it will be put at the beginning of the diaspora* post.

**Comment by manu-p** [comment]
Thanks for the explanation noplanman! Very clear, I'll put in on and check it out shortly. Thanks again. :)

**Comment by manu-p** [comment]
Hi @noplanman, it did not work : nothing was posted in diaspora* after inserting the code in function.php (child theme), while it'd been working before, i.e. without the code.

Where could I find some logging to investigate?

**Comment by noplanman** [comment]
What version of PHP are you using? I'm guessing 5.3.

Try this instead:
```php
function wp2d_title_with_featured_image( $default, $title, $permalink ) {
	return $default . '<br />' . get_the_post_thumbnail();
}
add_filter( 'wp2d_title_filter', 'wp2d_title_with_featured_image', 10, 3 );
```

**Comment by manu-p** [comment]
This site uses PHP7.
I'll check with this new code and tell you. Thanks again.

**Comment by manu-p** [comment]
So here's what's been done and what I got:

The post just holds a few words and a photo with a caption. I expect to find the text and the photo reposted in D*.
Instead, I find the "_few words_" and _[caption id=“attachment_2768” align=“aligncenter” width=“640”] the caption[/caption]_

Originally posted at: _blabla_

**Comment by noplanman** [comment]
This is very strange, not sure why it's not working.

If possible, give me the following info so I can explore this issue a bit:
- WordPress Version
- WordPress Theme
- diaspora* pod

**Comment by manu-p** [comment]
WP version : 4.6.1
WP theme : surreal
D* pod : diasp.org

Now, I found the reason why I got no picture posted in D* : it's the **featured-image** that's used in the D* post. Until now, I did not use this feature since it was not useful for the blog and the mailpoet newsletter. I'll have to check it with the next newsletter, but the point is now solved with wp-to-diaspora!

Thanks @noplanman!

**Comment by noplanman** [comment]
Ok, thought you had set that, should have asked to make sure 😄 

Anyway, glad to hear all is good now!

**Comment by manu-p** [comment]
Hi @noplanman, I'm back with this problem with a new question.

I can see the "featured-image" displayed in D* when I publish a post "by hand", i.e. right now, **not** when I choose to **schedule the publication** at another date/time.

**Comment by noplanman** [comment]
@manu-p Thanks for reporting this, I can reproduce it on my installation.

Will look into it 👍

**Comment by noplanman** [comment]
Right, the problem seems to be the context here. When publishing directly, the `$post` variable is set by the admin panel, but when scheduled, it isn't in the same way.

@gutobenn The best option would be to alter the filter itself, to add the post ID as a parameter, kind of how most filters work.
Question is how we should implement this to keep breakage to a minimum, if this is even a necessary consideration.

Most optimal would be to change the current filter, meaning that all setups using the current filter as it is now, will stop working properly. From the posts I see on diaspora*, I think most people don't make use of the filter, so changing it shouldn't make such a huge impact.

Maybe even something as simple as this (just 2 parameters, WP2D_Post object and the default value):
```php
add_filter( 'wp2d_title_filter',
	function( $wp2d_post, $default ) {
		return $default . '<br />' . get_the_post_thumbnail( $wp2d_post->ID );
	}, 10
);
```

**Comment by gutobenn** [comment]
@noplanman I think we can do that

**Comment by noplanman** [comment]
@manu-p Have updated the plugin to 1.8.0 now, which has the updated filter.

This allows you to use the following, which works for instant and scheduled posts:
```php
add_filter( 'wp2d_title_filter',
	function( $wp2d_post, $default ) {
		return $default . '<br />' . get_the_post_thumbnail( $wp2d_post->ID );
	}, 10, 2
);
```

**Comment by manu-p** [comment]
Works like a charm! Thanks!

**Comment by noplanman** [comment]
@manu-p Please not I've modified the example above, as it had a little bug in it, the order of the parameters have been switched.

Also, make sure you're using 1.9.1!! Prior versions have this bug which slipped through 😖

**Comment by manu-p** [comment]
OK thanks, corrected.  ;-)

**Comment by manu-p** [comment]
You know we, users, always want more, don't you!?

What can I do to get the photo caption in WP exporter to D* ? I mean, so far the WP post featured-image is properly displayed in D*, but not its caption... That would be nice sometimes, don't you agree ?

**Comment by noplanman** [comment]
> You know we, users, always want more, don't you!?

Good! 😃👍

This one is a bit more involved, but pretty easy too:
```php
add_filter( 'wp2d_title_filter',
	function ( $default, $wp2d_post ) {
		/** @var WP2D_Post $wp2d_post */
		$caption = $wp2d_post->custom_img_caption( '', [
			'caption' => get_the_post_thumbnail_caption( $wp2d_post->ID ),
		], get_the_post_thumbnail( $wp2d_post->ID ) );

		return $default . '<br />' . $caption;
	}, 10, 2
);
```

If you need any explanation, just let me know and I'll happily elaborate.

**Comment by manu-p** [comment]
OK thanks @noplanman, I tell you later with the next post.

**Comment by manu-p** [comment]
And the winner is... **YOU**! :+1: 
Thanks again @noplanman
